[server] For GitLab projects without an owner avatar, fall back to the namespace avatar, or generate the default GitLab avatar

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

For GitLab projects belonging to a group, it seems that there is no `owner.avatar_url` to be found here:

https://github.com/gitpod-io/gitpod/blob/fc68e4780eae27c5a3e7c0907f457ebb393a0371/components/server/ee/src/gitlab/gitlab-app-support.ts#L46

Example group project from GitLab's API:

```ts
{
  "id": 23116976,
  "description": "",
  "name": "sample-project",
  "name_with_namespace": "reproduction-group / sample-project",
  /* ... (no "owner") ... */
  "namespace": {
    "id": 10408914,
    "name": "reproduction-group",
    /* ... */
    "avatar_url": "/uploads/-/system/group/avatar/10408914/gitpod-ua.png",
    "web_url": "https://gitlab.com/groups/reproduction-group"
  },
  /* ... */
}
```

As you can see, there is no `"owner"`, but there is an avatar URL on the `"namespace"` (although it's relative).

When there is no owner avatar URL, this Pull Request falls back to the namespace avatar URL if available.

EDIT: When no avatar is available, it generates a SVG avatar that looks like GitLab's own default group avatar.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/5330

## How to test
<!-- Provide steps to test this PR -->

1. Connect with GitLab
2. Be part of a group which has an avatar (e.g. https://gitlab.com/reproduction-group)
3. Go to `/new` -- the GitLab group should have the set avatar visible

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
[server] For GitLab projects without an owner avatar, fall back to the namespace avatar, or generate the default GitLab avatar
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

andrew-farries commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +61,18 @@ export class GitLabAppSupport {
         return result;
     }
+
+    protected async getAccountAvatarUrl(anyProject: any, providerHost: string): Promise<string> {
```
Is it possible to do any better than an `any` type here?

jankeromnes replied:
Good catch!

The original code uses `const anyProject = project as any`, so I've just used that in this method.

It looks like the types returned by the API are of type `ProjectSchema`, but also that type seems incomplete:

<img width="957" alt="Screenshot 2022-03-15 at 14 40 01" src="https://user-images.githubusercontent.com/599268/158390455-e603b461-95be-4f8f-a09a-5a1214e6c156.png">

jankeromnes replied:
We could potentially reverse-engineer the type, e.g.:

```ts
type ProjectSchema = {
    owner?: {
        avatar_url: string;
    },
    namespace: {
        avatar_url: string;
    },
};
```

but I'm not sure if this is better than `any`. 🤔

andrew-farries replied:
One option would be to get the data returned by the API and then use `typeof` to generate a type from it, if there is no 'official' specification available.

gtsiolis commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +61,18 @@ export class GitLabAppSupport {
+            return anyProject.owner.avatar_url;
+        }
+        if (anyProject.namespace?.avatar_url) {
+            let url = anyProject.namespace.avatar_url;
```
**issue**: Not sure if this is the correct line to comment this but the avatar seems broken now while before it was just a white blank space, see screenshots below.

| BEFORE |  AFTER|
|-|-|
| <img width="346" alt="list-before" src="https://user-images.githubusercontent.com/120486/158410173-fd1b2e71-6f31-40b0-b2af-22dc7c8577af.png"> | <img width="346" alt="list-after" src="https://user-images.githubusercontent.com/120486/158410180-58699953-f535-4934-9cd7-6a7a356c0fc9.png"> |

jankeromnes replied:
Ah, thanks for catching this @gtsiolis. If you inspect the broken icons, what does the avatar URL look like?

gtsiolis replied:
**praise**: Thanks for fixing this, work like a charm. ✨

gtsiolis replied:
The groups with custom avatars still seems broken. What do you think of opening a follow up issue for this to keep track of this?

<img width="489" alt="Frame 370" src="https://user-images.githubusercontent.com/120486/161529256-f313f045-110a-47ff-afba-1bb4647e8737.png">

gtsiolis replied:
It seems to be working ok now for all but one of the groups in there, posted a comment in https://github.com/gitpod-io/gitpod/issues/9112#issuecomment-1087787341.

gtsiolis reviewed (commented):
Thanks for picking this up, @jankeromnes!

Commented below about two outstanding issues related to UX here.

gtsiolis commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +61,18 @@ export class GitLabAppSupport {
     }
+
+    protected async getAccountAvatarUrl(anyProject: any, providerHost: string): Promise<string> {
+        if (anyProject.owner?.avatar_url) {
```
**thought**: Alternatively, we could introduce a custom fallback image using GitLab's logo for the groups that we cannot fetch their custom generated avatar, see [relevant comment](https://github.com/gitpod-io/gitpod/issues/5330#issuecomment-1067954451):

> GitLab auto-generates an avatar based on the group name and using a colored background in combination with the first letter of the group name.

jankeromnes replied:
Not known. Would you know how I can become part of such a group with less permissions, in order to reproduce this?

Alternatively, I can add debug logs to this PR, and ask you to load that page again. 😊

jankeromnes replied:
Also, I agree that the fallback image would be nice, but [I couldn't find how to generate it](https://github.com/gitpod-io/gitpod/pull/8824#issuecomment-1067956035):

> Note: When there is no avatar, it would be nice to use the same default avatar as GitLab (i.e. a big letter with a pastel background). However, I couldn't find how to generate this, so for now I left the avatar URL empty when there is no avatar.

Would you know a potential way to find out? 👀

gtsiolis replied:
@jankeromnes Last time I remember contributing to this[[1](https://gitlab.com/gitlab-org/gitlab-foss/-/merge_requests/17271)], GitLab was using a custom logic to create the icon, see [relevant code](https://gitlab.com/gitlab-org/gitlab/-/blob/b2a22b6e85200ce55ab09b5c765043441b086c96/app/helpers/avatars_helper.rb#L151-161). Maybe this is why this is not available through their API?

jankeromnes replied:
Ah, thank you so much @gtsiolis for linking the relevant code ❤️ I couldn't find it, and now I've used it to implement a similar defafult-avatar-generator for our GitLab avatar fetching code.

jankeromnes commented:
Ready for another round of reviews! 🚀

I've implemented a default avatar generator similar to GitLab's one (thanks @gtsiolis for linking the relevant code! 🎯) so there should always be an avatar now.

I've also implemented more specific types (instead of using `any`), based on the official types from our library (`gitbeaker`) with the additional missing fields.

gtsiolis commented:
Taking another look now! 👀

gtsiolis reviewed (commented):
Took another look and left some more comments below. Thanks, @jankeromnes! 🏀

gtsiolis commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +75,27 @@ export class GitLabAppSupport {
+        // - https://gitlab.com/gitlab-org/gitlab-foss/-/blob/84b4743475246e91dc78c3f25f9b335c40be84cd/app/assets/stylesheets/startup/startup-general.scss#L420-422
+        const text = owner.name[0].toUpperCase();
+        const backgroundColor = ["#fcf1ef", "#f4f0ff", "#f1f1ff", "#e9f3fc", "#ecf4ee", "#fdf1dd", "#f0f0f0"][
+            owner.id % 7
```
**issue**: One minor issue here is that some avatars are using a different background color from what GitLab does for each group. I cound not reproduce this for new groups but some of the existing groups. Is this expected?

| on GitLab.com | on Gitpod.io |
|-|-|
| <img width="384" alt="Screenshot 2022-03-28 at 10 12 54 PM" src="https://user-images.githubusercontent.com/120486/160470100-316716e6-df23-4eff-be64-3f798d35d27b.png"> | <img width="637" alt="Screenshot 2022-03-28 at 10 13 06 PM" src="https://user-images.githubusercontent.com/120486/160470101-8ef38242-b1b4-4270-b662-6ac59301d628.png"> |

jankeromnes replied:
Interesting! Would have to look at debug logs -- maybe we're not using the right ID for these groups 👀

gtsiolis replied:
Yeah, probably also minor issue that's not worth blocking us from merging this. 💭

jankeromnes replied:
I think this might also be a case of "using the avatar of a nested group": https://github.com/gitpod-io/gitpod/pull/8824#discussion_r839377129

gtsiolis replied:
This seems to be also resolved now!

jankeromnes commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +75,27 @@ export class GitLabAppSupport {
+            owner.id % 7
+        ];
+        const svg = `<svg viewBox="0 0 32 32" height="32" width="32" style="background-color: ${backgroundColor}" xmlns="http://www.w3.org/2000/svg">
+            <text x="50%" y="50%" dominant-baseline="middle" text-anchor="middle" style='font-size: 0.875rem; font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Noto Sans", Ubuntu, Cantarell, "Helvetica Neue", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"'>
```
They are supposed to be centered. I'll debug the produced SVG. 🔍

jankeromnes commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +75,27 @@ export class GitLabAppSupport {
+        // If there is no avatar, generate the same default avatar that GitLab uses. Based on:
+        // - https://gitlab.com/gitlab-org/gitlab/-/blob/b2a22b6e85200ce55ab09b5c765043441b086c96/app/helpers/avatars_helper.rb#L151-161
+        // - https://gitlab.com/gitlab-org/gitlab-foss/-/blob/84b4743475246e91dc78c3f25f9b335c40be84cd/app/assets/stylesheets/startup/startup-general.scss#L1611-1631
+        // - https://gitlab.com/gitlab-org/gitlab-foss/-/blob/84b4743475246e91dc78c3f25f9b335c40be84cd/app/assets/stylesheets/startup/startup-general.scss#L420-422
```
Good point. I guess we could somehow re-use the same fonts as in our dashboard, but:
- Since they're not exactly the same as GitLab, I'm worries that in some cases they might look different than GitLab icons (which fails to achieve our goal here)
- Also, I'm not entirely sure how to reference the Tailwind class from inside the SVG (which doesn't use Tailwind) 🤔

So, maybe copying the fonts from GitLab just for the GitLab default avatar makes sense? 😇

gtsiolis replied:
To clarify, the question aimed to suggest not importing any fonts but relying on our font stack to render the character inside the group avatar. We do use the same background color which already helps, right?

jankeromnes replied:
This is what the SVG looks like for me when I remove the font specification:

<img width="158" alt="Screenshot 2022-03-30 at 15 49 22" src="https://user-images.githubusercontent.com/599268/160850253-edcc733c-7d50-4944-a5db-2a4595a1d93b.png">

gtsiolis replied:
Oh, I meant rely on (use) our font stack instead. 😇

jankeromnes commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +75,27 @@ export class GitLabAppSupport {
+        }
+        // If there is no avatar, generate the same default avatar that GitLab uses. Based on:
+        // - https://gitlab.com/gitlab-org/gitlab/-/blob/b2a22b6e85200ce55ab09b5c765043441b086c96/app/helpers/avatars_helper.rb#L151-161
+        // - https://gitlab.com/gitlab-org/gitlab-foss/-/blob/84b4743475246e91dc78c3f25f9b335c40be84cd/app/assets/stylesheets/startup/startup-general.scss#L1611-1631
```
Not sure what the difference is. I'll try to find equivalent sources from the `gitlab` repo.

gtsiolis replied:
> Not sure what the difference is.
The difference is that gitlab-foss does not contain any enterprise or licensed code, mostly everything inside the `ee/` top folder which is automatically dropped. Some open source orgs have a requirement of not using enterprise or licenced code and strictly use GitLab's FOSS version.

I've already linked the equivalent source on the `gitlab` repository in the code suggestion in https://github.com/gitpod-io/gitpod/pull/8824#discussion_r836723888.

jankeromnes replied:
Haha, since GitHub was being slow, it took me 5 minutes to understand that you actually linked to the comment right above mine 😅 fair enough, many thanks for finding the correct source!

jankeromnes commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -9,6 +9,20 @@ import { inject, injectable } from "inversify";
+type ProjectSchema = ProjectSchemaDefault & {
+    last_activity_at: string;
+    namespace: NamespaceInfoSchemaDefault & {
+        avatar_url: string;
```
Good catch! Found [a way to support emojis](https://dev.to/acanimal/how-to-slice-or-get-symbols-from-a-unicode-string-with-emojis-in-javascript-lets-learn-how-javascript-represent-strings-h3a) as first character. 👍

jankeromnes replied:
@gtsiolis Could you please try again, or invite me to that group? (For some reason, I'm not able to create one that starts with an emoji)

gtsiolis replied:
@jankeromnes It almost worked! I added this to the reproduction-group in GitLab (see [example project](https://gitlab.com/reproduction-group/sample-group/sample-project)), but somehow the emoji is surfacing to the top-level group. Previously, refrained from adding this as it would break new project wizard for you, too. 🤷  

<img width="487" alt="Screenshot 2022-03-30 at 7 01 12 PM" src="https://user-images.githubusercontent.com/120486/160901253-ee7cebe1-4aa5-4666-b1c6-41f509db2a09.png">

jankeromnes replied:
@gtsiolis Ah, I think I got it:
- Previously, we picked the avatar only for the `owner` (when provided by GitLab)
- With this Pull Request, we also look at the `namespace` when the `owner` is not available
- However, I see that `namespace` is not always the "root namespace" (i.e. for `reproduction-group / 🍊-sample-group / 🍋-sample-project`, GitLab provides `🍊-sample-group`, not `reproduction-group`)

I believe the fix is to only use the avatar for `owner` if provided, or `namespace` if it is the root namespace. (Note: We could also look at the `parent_id` of nested groups, then make follow-up GitLab API requests to find the root namespace and its avatar -- but I think that would be way over-engineered at this point.)

Let's see how the new algorithm behaves. Pushing soon.

gtsiolis replied:
Looks resolved now. Thanks for fixing this, @jankeromnes!

jankeromnes commented:
Briefly taking this back to Draft for some quick debugging.

jankeromnes commented:
_Phew._ Okay now this should be ready to rock'n'roll. 🎸😁

@gtsiolis please take another look when you have some time. 🙏

(For what it's worth, I initially estimated this work as a 0.25 day effort / 1 iteration -- looks like I was about 10x off on both. 😅 )

easyCZ commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +76,36 @@ export class GitLabAppSupport {
+        // - https://gitlab.com/gitlab-org/gitlab/-/blob/b2a22b6e85200ce55ab09b5c765043441b086c96/app/helpers/avatars_helper.rb#L151-161
+        // - https://gitlab.com/gitlab-org/gitlab/-/blob/861f52858a1db07bdb122fe947dec9b0a09ce807/app/assets/stylesheets/startup/startup-general.scss#L1611-1631
+        // - https://gitlab.com/gitlab-org/gitlab/-/blob/861f52858a1db07bdb122fe947dec9b0a09ce807/app/assets/stylesheets/startup/startup-general.scss#L420-422
+        const backgroundColor = ["#fcf1ef", "#f4f0ff", "#f1f1ff", "#e9f3fc", "#ecf4ee", "#fdf1dd", "#f0f0f0"][
+            owner.id % 7
```
```suggestion
        const backgroundColors = ["#fcf1ef", "#f4f0ff", "#f1f1ff", "#e9f3fc", "#ecf4ee", "#fdf1dd", "#f0f0f0"];
        const backgroundColor = backgroundColors[owner.id % backgroundColors.length]
```
To avoid the future fix-up when a new color is added to the mix.

easyCZ commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -237,7 +237,11 @@ export default function NewProject() {
     const accounts = new Map<string, { avatarUrl: string }>();
     reposInAccounts.forEach((r) => {
-        if (!accounts.has(r.account)) accounts.set(r.account, { avatarUrl: r.accountAvatarUrl });
+        if (!accounts.has(r.account)) {
```
👍 Much nicer on the eyes

easyCZ reviewed (approved):
Code-wise looks good to me. Adding hold to resolve remaining questions.

/hold

gtsiolis commented:
Taking another look at this now! 👀

gtsiolis commented on components/server/ee/src/gitlab/gitlab-app-support.ts:
```diff
@@ -61,4 +76,36 @@ export class GitLabAppSupport {
+        const backgroundColor = ["#fcf1ef", "#f4f0ff", "#f1f1ff", "#e9f3fc", "#ecf4ee", "#fdf1dd", "#f0f0f0"][
+            owner.id % 7
+        ];
+        // Uppercase first character of the name, support emojis, default to whitespace.
```
**praise**: Lovely comment! 💯

gtsiolis reviewed (approved):
LGTM, @jankeromnes! 🔮

jankeromnes commented:
Nits addressed & follow-ups filed! 🎉

@easyCZ or @gtsiolis could you please re-approve?

/unhold

gtsiolis reviewed (approved):
Works like a charm, @jankeromnes! 🔮 

Thanks for also opening the follow up issue in https://github.com/gitpod-io/gitpod/pull/8824#discussion_r841945398 to track one potentially minor edge case. 💯 

Approving for UX, but you'll still need an approval from someone from the WebApp team.
> Waiting on code owner review from gitpod-io/engineering-webapp.
